lnwallet/channel.go

```diff
@@ -1959,14 +1965,18 @@ func (lc *LightningChannel) restorePendingLocalUpdates(
 					lc.localUpdateLog.htlcCounter))
 			}
 
-			lc.localUpdateLog.appendHtlc(payDesc)
-
+			if lc.localUpdateLog.restoreHtlc(payDesc) {
+				lc.localUpdateLog.htlcCounter++
```

Comment:
alternatively, should these increments happen in `restoreHtlc`/`restoreUpdate`? i think it would reduce the chance of error in having to remember to do the increments manually

Reply:
Yeah I did that at first. The update or htlc may already be counted for in the update log (via indices), but just not be in the update log. In this case, we need to restore and not increment to prevent a de-sync of the update logs.

Reply:
When the update logs are initialized, they are [created from the local commitment and the remote commitment](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/channel.go#L1319). So depending on which it came from, the logindex and htlccounter are already set. If the update or htlc we are restoring is already on the source commitment, then there's no need to increment. For example, incrementing [here](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/channel.go#L1810) would be incorrect behavior as this update is already accounted for in the indices of the update log.

Reply:
I do agree that this custom logic is rather confusing and that custom logic generally leads to more bugs. I was thinking of refactoring the update logs so that they were a little more intuitive to use. Or maybe some bigger overhaul of the channel state machine.

Reply:
Could use a comment here at the very least explaining why we need to manually increment it. Or better yet extract this logic into a new method on the update log itself.

Reply:
Discussed offline with conner, going to have an argument to restoreHtlc / Update and also add a comment

Reply:
Realized we don't need this. Double-checked all the places that we restore from and we should be fine without this. When restoring with `appendHtlc` or `appendUpdate` we just have to be sure that the order of updates before *and* after are the same as otherwise htlcs/updates will be in the wrong order.